jobhub.go

```diff
@@ -0,0 +1,190 @@
+		p.id = nextIDPipeline()
+	}
+	job.pipelineID = p.id
+	job.id = p.nextIDJob()
+	p.jobContainer = append(p.jobContainer, job)
+	p.jobByID[job.id] = job
+	p.startingJob[job.id] = true
+	return job
+}
+
+func (p *Pipeline) AddJobDependency(job Job, deps ...Job) {
+	var tempContainer []Job
+	tempContainer = append(deps, job)
+	if p.id != 0 {
+		if p.jobContainer != nil {
+			for _, givenJob := range tempContainer {
+				if givenJob.id == p.id {
+					continue
+				} else {
+					p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] does not belong to this pipeline",
+						p.id, p.Name, givenJob.id, givenJob.Name)
+				}
+			}
+			for _, givenJob := range tempContainer {
+				for _, addedJob := range p.jobContainer {
+					if addedJob == givenJob {
+						continue
+					} else {
+						p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] has never been added",
+							p.id, p.Name, givenJob.id, givenJob.Name)
+					}
+				}
+			}
+			for _, d := range deps {
+				p.jobDependency[job.id] = append(p.jobDependency[job.id], d.id)
+				delete(p.startingJob, d.id)
+			}
+		} else {
+			p.Log.Panicf("Pipeline [%d][%s] | Job Container is empty",
+				p.id, p.Name)
+		}
+	} else {
+		p.Log.Panicf("Pipeline [%d][%s] | Pipeline not initialized",
+			p.id, p.Name)
+	}
+}
+
+func (p *Pipeline) resolveDependencyRecursion(jobID int, level int) {
+	if l := p.recursionLevels[jobID]; l >= level {
+		return
+	}
+	p.recursionLevels[jobID] = level
+	level++
+	for _, depID := range p.jobDependency[jobID] {
+		p.Log.Debugf("%s -> %s | recursion level: %d", p.jobByID[jobID], p.jobByID[depID], level)
+		p.resolveDependencyRecursion(depID, level)
+	}
+}
+
+func (p *Pipeline) resolveDependency() []int {
+	var queue []int
+	tempRL := make(map[int]int)
+	for startID, _ := range p.startingJob {
+		p.resolveDependencyRecursion(startID, 1)
+	}
+	for jID, l := range p.recursionLevels {
+		tempRL[jID] = l
+	}
+	for i := 0; i < len(tempRL); {
+		var jobID int
+		max := 0
+		for jID, l := range tempRL {
+			if l > max {
+				max = l
+				jobID = jID
+			}
+		}
+		queue = append(queue, jobID)
+		delete(tempRL, jobID)
+	}
+	return queue
+}
+
+func (p *Pipeline) runJob(job Job) (*exitStatus, error) {
+	_, err := os.Stat(job.Path)
+	if err != nil {
+		return nil, err
+	}
+	process := exec.Command(job.Path)
+	start := time.Now()
+	err = process.Run()
+	elapsed := time.Since(start)
+	exitError, ok := err.(*exec.ExitError)
+	if !ok {
+		p.Log.Panicf("Cannot cast to exitError", err)
+	}
+	return &exitStatus{runtime: elapsed, status: exitError.Sys().(syscall.WaitStatus)}, err
+}
+
+func (p *Pipeline) Run() {
+	if p.jobDependency != nil {
+		queue := p.resolveDependency()
+		for _, jID := range queue {
+			exitStatus, err := p.runJob(p.jobByID[jID])
+			if err != nil {
+				p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s][t: %f] | %s",
+					p.id, p.Name, jID, p.jobByID[jID].Name, exitStatus.runtime, err)
+			}
+		}
+	} else {
+		p.Log.Panicf("Pipeline [%d][%s] | No jobs in queue", p.id, p.Name)
```

Comment:
We may have jobs which do not depend on each other, we still want to run them. ;-)